contents/data/nosql/NoSQL.md

```diff
@@ -0,0 +1,371 @@
+
+Some common implementations of such databases are key-value stores, document stores (which are an extension of key-value stores),
+and though less common - graph databases (which can also be considered an extension of document stores).
+
+---
+
+## How does NoSQL work?
+
+One such commonly used database is MongoDB, which is a document based database.
+
+### Basic structure of document-based databases
+
+![](documentDatabaseStructure.png)
+
+In contrast to tables and table entries in SQL databases, we typically have collections and documents.
+A database comprises of multiple collections, which in turn consists of multiple documents.
+
+In a simplistic e-commerce website for example, you may have the following collections:
+- customers - storing the account details of customers, their purchase histories, etc.
+- items - a collection of all items (documents) available for purchase
+- admin - a collection storing admin account details
+- **...**
+
+The following might be how an item available for purchase under the *items* collection
+is represented in a JSON format.
+
+
+```js
+{
+  type: "book",
+  price: 20,
+  popularity: 9.7,
+  ...
+}
+```
+
+<br>
+<box type="info" light>
+Note that while many NoSQL databases provide a JSON interface to interact with the data,
+the underlying storage implementation may be different for performance reasons.<br>
+For example, MongoDB stores documents in BSON (json encoded in binary). 
+</box>
+
+<br>
+
+### Basic CRUD operation examples
+
+Let's skip the trivial details of database setup which would differ across different database technologies,
+and look at how you would interact with NoSQL databases.
+
+Unlike SQL, where database interaction is achieved through the different dialects of the query language,
+most interaction with NoSQL databases is achieved in a simple and intuitive _object oriented manner_,
+and JSON-like queries.
+
+Let's get back to the above example. Say you already have the _customers_ collection set up and ready to go.
+A potential new user sends a POST request to register a new account, and your server would then simply call
+the following mongoDB code to store the details after some processing.
+
+```js
+db.customers.insertOne({
+  username: "panda",
+  password: hashedPassword,
+  email: "panda@pandas.com"
+})
+```
+
+That was a rather simple use case, but you get the idea. Let's get into something more powerful.
+
+<p id="schemaless">
+Suppose our _items_ collection is structured like so:
+</p>
+
+```js
+[
+  {
+    type: "book",
+	title: "about pandas"
+    price: 20,
+    popularity: 9.7,
+	author: "panda1"
+    ...
+  },
+  {
+	title: "more pandas",
+	author: "panda2",
+  },
+  ...
+]
+```
+
+Here, the code simply provides a query JSON object matching the structure of the items collection.
+Specifically, items with a price of less than 30 are filtered with the `$lt: 30` key-value pair,
+and items with a popularity of more than 8 are filtered with the `$gt: 8`.
```

Comment:
Is the above code block the json object in response to a query? If yes, It would help if there was a sample of the query as well.

Reply:
I discarded an entire chunk above by accident 😅